core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java

```diff
@@ -42,7 +42,13 @@ public void serialize(LongDecimalWithOverflowAndLongState state, BlockBuilder ou
             long overflow = state.getOverflow();
             long[] decimal = state.getDecimalArray();
             int offset = state.getDecimalArrayOffset();
-            VARBINARY.writeSlice(out, Slices.wrappedLongArray(count, overflow, decimal[offset], decimal[offset + 1]));
+            if (count == 1 && overflow == 0 && decimal[offset] == 0) {
```

Comment:
`count == 1 && overflow == 0` should be separate from `decimal[offset] == 0`

`decimal[offset] == 0` will be true often when `count == 1 && overflow == 0` isn't.

So you have cases:
* 1 long -> count == 1 && overflow == 0 && decimal[offset] == 0
* 2 longs -> count == 1 && overflow == 0
* 3 longs -> decimal[offset] == 0
* 4 longs -> full case

I think you can even make it branchless:
* append `decimal[offset + 1]` unconditionally, len += 1
* append `decimal[offset]` unconditionally, len = len + (decimal[offset] == 0 ? 0 : 1)
* append  `overflow` and `count`, len = len + ((count == 1 & overflow == 0) ? 0 : 2)

Reply:
this may help in some cases, I will try that. This probably won't help in tpch q17/ q18 as it's already hitting the condition.